File: components/api/api-modules-python/src/main/resources/META-INF/dirigible/python-modules/dirigible/core/destinations.py
# Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License v2.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-v20.html
# SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
# SPDX-License-Identifier: EPL-2.0

from org.eclipse.dirigible.components.api.core import DestinationsFacade
from org.eclipse.dirigible.components.api.cf import CloudFoundryModule
from org.eclipse.dirigible.components.api.kyma import KymaModule
from org.eclipse.dirigible.http.client import httpClient
from java.lang import Class
import json

def get(name):
    if is_cloud_environment():
        return json.loads(get_cloud_destination(name))
    return json.loads(DestinationsFacade.get(name))

def set(name, destination):
    if is_cloud_environment():
        create_or_update_cloud_destination(name, destination)
    DestinationsFacade.set(name, json.dumps(destination))

def remove(name):
    if not is_cloud_environment():
        raise Exception("The delete destination operation is not supported for non-cloud environments")
    delete_cloud_destination(name)

def is_cloud_environment():
    try:
        Class.forName("org.eclipse.dirigible.cf.CloudFoundryModule")
        return True
    except:
        pass

    try:
        Class.forName("org.eclipse.dirigible.kyma.KymaModule")
        return True
    except:
        pass

    return False

def get_cloud_destination(name):
    token = get_oauth_token()
    destination_url = f"{get_destinations_base_path()}/{name}"
    headers = [
        {"name": "Authorization", "value": f"{token.token_type} {token.access_token}"}
    ]
    response = httpClient.get(destination_url, {"headers": headers})

    if response.statusCode == 404:
        raise Exception(f"Destination with name '{name}' not found")

    return response.text

def get_destinations_base_path():
    return f"{configurations.get('DIRIGIBLE_DESTINATION_URI')}/destination-configuration/v1/destinations"

def get_instance_destinations_base_path():
    return f"{configurations.get('DIRIGIBLE_DESTINATION_URI')}/destination-configuration/v1/instanceDestinations"

def create_or_update_cloud_destination(name, destination):
    is_existing_destination = True
    try:
        get_cloud_destination(name)
    except:
        is_existing_destination = False

    instance_destination_url = get_instance_destinations_base_path()
    destination["Name"] = name
    token = get_oauth_token()
    headers = [
        {"name": "Authorization", "value": f"{token.token_type} {token.access_token}"}
    ]
    options = {
        "headers": headers,
        "text": json.dumps(destination)
    }

    if is_existing_destination:
        response = httpClient.put(instance_destination_url, options)
        if response.statusCode != 200:
            raise Exception(f"Error occurred while updating destination '{name}': {response.text}")
    else:
        response = httpClient.post(instance_destination_url, options)
        if response.statusCode != 201:
            raise Exception(f"Error occurred while creating destination '{name}': {response.text}")

def delete_cloud_destination(name):
    token = get_oauth_token()
    headers = [
        {"name": "Authorization", "value": f"{token.token_type} {token.access_token}"}
    ]
    url = f"{get_instance_destinations_base_path()}/{name}"
    response = httpClient.delete(url, {"headers": headers})
    if response.statusCode != 200:
        raise Exception(f"Error occurred while deleting destination '{name}': {response.text}")

def get_oauth_token():
    oauth_client_id = configurations.get("DIRIGIBLE_DESTINATION_CLIENT_ID")
    oauth_client_secret = configurations.get("DIRIGIBLE_DESTINATION_CLIENT_SECRET")
    oauth_url = configurations.get("DIRIGIBLE_DESTINATION_URL")
    uri = configurations.get("DIRIGIBLE_DESTINATION_URI")

    if not oauth_client_id or not oauth_client_secret or not oauth_url or not uri:
        raise Exception("Invalid destination configuration")

    oauth_config = {
        "url": oauth_url,
        "clientId": oauth_client_id,
        "clientSecret": oauth_client_secret
    }

    client = oauth.getClient(oauth_config)
    return client.getToken()

File: components/api/api-modules-javascript/src/main/resources/META-INF/dirigible/modules/src/core/destinations.ts
/*
 * Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
/**
 * API Destinations
 *
 * Note: This module is supported only with the Mozilla Rhino engine
 */

import { configurations } from ".";
import { httpClient } from "@dirigible/http/client";
let oauth = dirigibleRequire("security/oauth"); // TODO: check this API

const JClass = Java.type("java.lang.Class");
const DestinationsFacade = Java.type("org.eclipse.dirigible.components.api.core.DestinationsFacade");

export class Destinations{

    public static get (name) {
        if (this.isCloudEnvironment()) {
            return JSON.parse(this.getCloudDestination(name));
        }
        return JSON.parse(DestinationsFacade.get(name));
    };

    public static set (name, destination) {
        if (this.isCloudEnvironment()) {
            this.createOrUpdateCloudDestination(name, destination);
        }
        DestinationsFacade.set(name, JSON.stringify(destination));
    };

    public static remove(name) {
        if (!this.isCloudEnvironment()) {
            throw new Error("The delete destination operation is not supported for non-cloud environments");
        }
        this.deleteCloudDestination(name);
    };

    public static isCloudEnvironment() {
        try {
            JClass.forName("org.eclipse.dirigible.cf.CloudFoundryModule");
            return true;
        } catch (e) {
            // Do nothing
        }
        try {
            JClass.forName("org.eclipse.dirigible.kyma.KymaModule");
            return true;
        } catch (e) {
            // Do nothing
        }
        return false;
    }

    private static getCloudDestination(name) {
        let token = this.getOAuthToken();

        let destinationUrl = `${this.getDestinationsBasePath()}/${name}`;

        let requestOptions = {
            headers: [{
                name: "Authorization",
                value: `${token.token_type} ${token.access_token}`
            }]
        };

        let response = httpClient.get(destinationUrl, requestOptions);

        if (response.statusCode === 404) {
            throw new Error(`Destination with name '${name}' not found`);
        };

        return response.text;

    }

    private static getDestinationsBasePath() {
        return `${configurations.get("DIRIGIBLE_DESTINATION_URI")}/destination-configuration/v1/destinations`;
    }

    private static getInstanceDestinationsBasePath() {
        return `${configurations.get("DIRIGIBLE_DESTINATION_URI")}/destination-configuration/v1/instanceDestinations`;
    }

    private static getSubaccountDestinationsBasePath() {
        return `${configurations.get("DIRIGIBLE_DESTINATION_URI")}/destination-configuration/v1/subaccountDestinations`;
    }

    private static createOrUpdateCloudDestination(name, destination) {
        let isExistingDestination = true;
        try {
            this.getCloudDestination(name);
        } catch (e) {
            isExistingDestination = false;
        }

        let instanceDestinationUrl = this.getInstanceDestinationsBasePath();
        destination.Name = name;

        let token = this.getOAuthToken();
        let requestOptions = {
            headers: [{
                name: "Authorization",
                value: `${token.token_type} ${token.access_token}`
            }],
            text: JSON.stringify(destination)
        };

        if (isExistingDestination) {
            let response = httpClient.put(instanceDestinationUrl, requestOptions);
            if (response.statusCode != 200) {
                throw new Error(`Error occurred while updating destinaton '${name}': ${response.text}`);
            }
        } else {
            let response = httpClient.post(instanceDestinationUrl, requestOptions);
            if (response.statusCode != 201) {
                throw new Error(`Error occurred while creating destinaton '${name}': ${response.text}`);
            }
        }
    }

    private static deleteCloudDestination(name) {
        let token = this.getOAuthToken();
        let requestOptions = {
            headers: [{
                name: "Authorization",
                value: `${token.token_type} ${token.access_token}`
            }]
        };
        let url = `${this.getInstanceDestinationsBasePath()}/${name}`;
        let response = httpClient.delete(url, requestOptions);
        if (response.statusCode != 200) {
            throw new Error(`Error occurred while deleting destinaton '${name}': ${response.text}`);
        }
    }

    private static getOAuthToken() {
        let oauthClientId = configurations.get("DIRIGIBLE_DESTINATION_CLIENT_ID");
        let oauthClientSecret = configurations.get("DIRIGIBLE_DESTINATION_CLIENT_SECRET");
        let oauthUrl = configurations.get("DIRIGIBLE_DESTINATION_URL");
        let uri = configurations.get("DIRIGIBLE_DESTINATION_URI");

        if (!oauthClientId || !oauthClientSecret || !oauthUrl || !uri) {
            throw new Error("Invalid destination configuration");
        }

        let oauthConfig = {
            url: oauthUrl,
            clientId: oauthClientId,
            clientSecret: oauthClientSecret
        };

        let client = oauth.getClient(oauthConfig);
        return client.getToken();
    }

}